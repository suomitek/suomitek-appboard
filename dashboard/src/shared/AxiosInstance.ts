import Axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";

import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  InternalServerError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

export function addAuthHeaders(axiosInstance: AxiosInstance) {
  axiosInstance.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
}

export function addErrorHandling(axiosInstance: AxiosInstance, store: Store<IStoreState>) {
  axiosInstance.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError = e;
      if (
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.expireSession());
      }
      let message = err.message;
      if (err.response) {
        if (err.response.data.message) {
          message = err.response.data.message;
        }
        if (typeof err.response.data === "string") {
          message = err.response.data;
        }
      }

      const dispatchErrorAndLogout = (m: string) => {
        // Global action dispatch to log the user out
        dispatch(actions.auth.authenticationError(m));
        // Expire the session if we are using OIDC tokens and
        // logout either way.
        dispatch(actions.auth.expireSession());
      };
      const response = err.response as AxiosResponse;
      switch (response && response.status) {
        case 401:
          dispatchErrorAndLogout(message);
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          // A 403 directly from the auth proxy requires reauthentication.
          if (Auth.usingOIDCToken() && Auth.is403FromAuthProxy(response)) {
            dispatchErrorAndLogout(message);
          }
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        case 500:
          return Promise.reject(new InternalServerError(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}

// Error handling is added with an interceptor in index.tsx
export const axios = Axios.create();
// Authorization headers and error handling are added with an interceptor in index.tsx
export const axiosWithAuth = Axios.create();
