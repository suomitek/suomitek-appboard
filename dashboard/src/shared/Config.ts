import axios from "axios";

interface ICluster {
  name: string;
  apiServiceURL: string;
  certificateAuthorityData?: string;
}

export interface IFeatureFlags {
  operators: boolean;
  ui: string;
}

// IConfig is the configuration for Kubeapps
export interface IConfig {
  namespace: string;
  appVersion: string;
  authProxyEnabled: boolean;
  oauthLoginURI: string;
  oauthLogoutURI: string;
  error?: Error;
  featureFlags: IFeatureFlags;
  clusters: ICluster[];
}

export default class Config {
  public static async getConfig() {
    const url = Config.APIEndpoint;
    const { data } = await axios.get<IConfig>(url);

    // Development environment config overrides
    // TODO(miguel) Rename env variable to KUBEAPPS_NAMESPACE once/if we eject create-react-app
    // Currently we are using REACT_APP_* because it's the only way to inject env variables in a sealed setup.
    // Please note that this env variable gets mapped in the run command in the package.json file
    if (process.env.NODE_ENV !== "production") {
      data.appVersion = "DEVEL";
      if (process.env.REACT_APP_KUBEAPPS_NS) {
        data.namespace = process.env.REACT_APP_KUBEAPPS_NS;
      }
    }

    return data;
  }

  private static APIEndpoint: string = "config.json";
}
