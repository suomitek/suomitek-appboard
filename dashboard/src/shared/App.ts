import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export const KUBEOPS_ROOT_URL = "api/tiller-deploy/v1";

export class App {
  public static async create(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.list(cluster, namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(
    cluster: string,
    namespace: string,
    releaseName: string,
    revision: number,
  ) {
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(
    cluster: string,
    namespace: string,
    releaseName: string,
    purge: boolean,
  ) {
    let endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  public static async listApps(cluster: string, namespace?: string, allStatuses?: boolean) {
    let endpoint = namespace
      ? url.kubeops.releases.list(cluster, namespace)
      : url.kubeops.releases.listAll(cluster);
    if (allStatuses) {
      endpoint += "?statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(endpoint);
    return data.data;
  }

  public static async getRelease(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      url.kubeops.releases.get(cluster, namespace, name),
    );
    return data.data;
  }
}
