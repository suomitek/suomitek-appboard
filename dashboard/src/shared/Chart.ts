import { JSONSchema4 } from "json-schema";
import { axiosWithAuth } from "./AxiosInstance";
import { IChart, IChartVersion } from "./types";
import * as URL from "./url";

export default class Chart {
  public static async fetchCharts(namespace: string, repo: string) {
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(
      URL.api.charts.list(namespace, repo),
    );
    return data.data;
  }

  public static async fetchChartVersions(namespace: string, id: string): Promise<IChartVersion[]> {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion[] }>(
      URL.api.charts.listVersions(namespace, id),
    );
    return data.data;
  }

  public static async getChartVersion(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion }>(
      URL.api.charts.getVersion(namespace, id, version),
    );
    return data.data;
  }

  public static async getReadme(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(
      URL.api.charts.getReadme(namespace, id, version),
    );
    return data;
  }

  public static async getValues(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(
      URL.api.charts.getValues(namespace, id, version),
    );
    return data;
  }

  public static async getSchema(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<JSONSchema4>(
      URL.api.charts.getSchema(namespace, id, version),
    );
    return data;
  }

  public static async listWithFilters(
    namespace: string,
    name: string,
    version: string,
    appVersion: string,
  ) {
    const url = `${
      Chart.APIEndpoint
    }/ns/${namespace}/charts?name=${name}&version=${encodeURIComponent(
      version,
    )}&appversion=${appVersion}`;
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(url);
    return data.data;
  }

  private static APIEndpoint: string = "api/assetsvc/v1";
}
