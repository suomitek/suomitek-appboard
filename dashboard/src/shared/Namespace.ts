import { axiosWithAuth } from "./AxiosInstance";
import * as url from "./url";

import { ForbiddenError, IResource, NotFoundError } from "./types";

export default class Namespace {
  public static async list(cluster: string) {
    // This call is hitting an actual backend endpoint (see pkg/http-handler.go)
    // while the other calls (create, get) are hitting the k8s API via the
    // frountend nginx.
    const { data } = await axiosWithAuth.get<{ namespaces: IResource[] }>(
      url.backend.namespaces.list(cluster),
    );
    return data;
  }

  public static async create(cluster: string, name: string) {
    const { data } = await axiosWithAuth.post<IResource>(url.api.k8s.namespaces(cluster), {
      apiVersion: "v1",
      kind: "Namespace",
      metadata: {
        name,
      },
    });
    return data;
  }

  public static async get(cluster: string, name: string) {
    try {
      const { data } = await axiosWithAuth.get<IResource>(url.api.k8s.namespace(cluster, name));
      return data;
    } catch (err) {
      switch (err.constructor) {
        case ForbiddenError:
          throw new ForbiddenError(
            `You don't have sufficient permissions to use the namespace ${name}`,
          );
        case NotFoundError:
          throw new NotFoundError(`Namespace ${name} not found. Create it before using it.`);
        default:
          throw err;
      }
    }
  }
}

// Set of namespaces used accross the applications as default and "all ns" placeholders
export const definedNamespaces = {
  all: "_all",
};
