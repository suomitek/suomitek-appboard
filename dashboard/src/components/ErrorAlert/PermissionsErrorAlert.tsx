import * as React from "react";
import { Lock } from "react-feather";

import { UnexpectedErrorAlert } from ".";
import { IRBACRole } from "../../shared/types";
import { namespaceText } from "./helpers";
import PermissionsListItem from "./PermissionsListItem";

interface IPermissionsErrorPage {
  action: string;
  roles: IRBACRole[];
  rawMessage: string;
  namespace: string;
}

class PermissionsErrorPage extends React.Component<IPermissionsErrorPage> {
  public render() {
    const { action, namespace, roles, rawMessage } = this.props;
    return (
      <UnexpectedErrorAlert
        title={
          <span>
            You don't have sufficient permissions to {action}{" "}
            {namespace && <>in {namespaceText(namespace)}</>}
          </span>
        }
        icon={Lock}
        showGenericMessage={false}
      >
        <div>
          {roles.length > 0 ? (
            <>
              <p>Ask your administrator for the following RBAC roles:</p>
              <ul className="error__permisions-list">
                {roles.map((r, i) => (
                  <PermissionsListItem key={i} namespace={namespace} role={r} />
                ))}
              </ul>
            </>
          ) : (
            <>
              <p>The following error was returned:</p>
              <div className="error__content">
                <section className="Terminal terminal__error elevation-1 type-color-white error__text">
                  {rawMessage}
                </section>
              </div>
            </>
          )}
          <p>
            See the documentation for more info on{" "}
            <a
              href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
              target="_blank"
              rel="noopener noreferrer"
            >
              access control in Kubeapps
            </a>
            .
          </p>
        </div>
      </UnexpectedErrorAlert>
    );
  }
}

export default PermissionsErrorPage;
