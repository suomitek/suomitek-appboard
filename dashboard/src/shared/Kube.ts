import * as url from "shared/url";
import { Auth } from "./Auth";
import { axiosWithAuth } from "./AxiosInstance";
import { ResourceKindsWithAPIVersions } from "./ResourceAPIVersion";
import { isNamespaced, ResourceKind, ResourceKindsWithPlurals } from "./ResourceKinds";
import { IK8sList, IResource } from "./types";

export const APIBase = "api/clusters/default";
export let WebSocketAPIBase: string;
if (window.location.protocol === "https:") {
  WebSocketAPIBase = `wss://${window.location.host}${window.location.pathname}`;
} else {
  WebSocketAPIBase = `ws://${window.location.host}${window.location.pathname}`;
}

// Kube is a lower-level class for interacting with the Kubernetes API. Use
// ResourceRef to interact with a single API resource rather than using Kube
// directly.
export class Kube {
  public static getResourceURL(
    cluster: string,
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    let u = `${url.api.k8s.base(cluster)}/${
      apiVersion === "v1" || !apiVersion ? "api/v1" : `apis/${apiVersion}`
    }`;
    if (namespace && isNamespaced(resource)) {
      u += `/namespaces/${namespace}`;
    }
    u += `/${resource}`;
    if (name) {
      u += `/${name}`;
    }
    if (query) {
      u += `?${query}`;
    }
    return u;
  }

  public static watchResourceURL(
    cluster: string,
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    let u = this.getResourceURL(cluster, apiVersion, resource, namespace);
    u = `${WebSocketAPIBase}${u}?watch=true`;
    if (name) {
      u += `&fieldSelector=metadata.name%3D${name}`;
    }
    if (query) {
      u += `&${query}`;
    }
    return u;
  }

  public static async getResource(
    cluster: string,
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    const { data } = await axiosWithAuth.get<IResource | IK8sList<IResource, {}>>(
      this.getResourceURL(cluster, apiVersion, resource, namespace, name, query),
    );
    return data;
  }

  // Opens and returns a WebSocket for the requested resource. Note: it is
  // important that this socket be properly closed when no longer needed. The
  // returned WebSocket can be attached to an event listener to read data from
  // the socket.
  public static watchResource(
    cluster: string,
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    return new WebSocket(
      this.watchResourceURL(cluster, apiVersion, resource, namespace, name, query),
      Auth.wsProtocols(),
    );
  }

  // Gets the plural form of the resource Kind for use in the resource path
  public static resourcePlural(kind: ResourceKind) {
    return ResourceKindsWithPlurals[kind];
  }

  // Gets the apiVersion of the resource Kind for use in the resource path
  public static resourceAPIVersion(kind: ResourceKind) {
    return ResourceKindsWithAPIVersions[kind];
  }
}
