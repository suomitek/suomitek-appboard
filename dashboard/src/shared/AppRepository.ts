import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";
import { ICreateAppRepositoryResponse } from "./types";
import * as url from "./url";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace));
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace, name));
    return data;
  }

  public static async resync(name: string, namespace: string) {
    const repo = await AppRepository.get(name, namespace);
    repo.spec.resyncRequests = repo.spec.resyncRequests || 0;
    repo.spec.resyncRequests++;
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(namespace, name), repo);
    return data;
  }

  public static async update(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.put<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.update(namespace, name),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(
      url.backend.apprepositories.delete(name, namespace),
    );
    return data;
  }

  // create uses the kubeapps backend API
  // TODO(mnelson) Update other endpoints to similarly use the backend API, removing the need
  // for direct k8s api access (for this resource, at least).
  public static async create(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.post<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.create(namespace),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async validate(repoURL: string, authHeader: string, customCA: string) {
    const { data } = await axiosWithAuth.post<any>(url.backend.apprepositories.validate(), {
      appRepository: { repoURL, authHeader, customCA },
    });
    return data;
  }

  private static APIBase: string = APIBase;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static getSelfLink(namespace: string, name?: string): string {
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories${
      name ? `/${name}` : ""
    }`;
  }
}
