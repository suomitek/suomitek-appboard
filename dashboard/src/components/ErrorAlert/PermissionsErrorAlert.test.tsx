import { shallow } from "enzyme";
import * as React from "react";

import { UnexpectedErrorAlert } from ".";
import { IRBACRole } from "../../shared/types";
import ErrorAlertHeader from "./ErrorAlertHeader";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import PermissionsListItem from "./PermissionsListItem";
import { genericMessage } from "./UnexpectedErrorAlert";

it("renders an error message for the action", () => {
  const roles: IRBACRole[] = [];
  const action = "unit-test";
  const wrapper = shallow(
    <PermissionsErrorAlert roles={roles} action={action} namespace="test" rawMessage="" />,
  );
  const header = wrapper
    .find(UnexpectedErrorAlert)
    .shallow()
    .find(ErrorAlertHeader);
  expect(header).toExist();
  expect(header.shallow().text()).toContain(`You don't have sufficient permissions to ${action}`);
  expect(wrapper.html()).toContain("The following error was returned:");
  expect(wrapper).toMatchSnapshot();
});

it("renders PermissionsListItem for each RBAC role", () => {
  const roles: IRBACRole[] = [
    {
      apiGroup: "test.kubeapps.com",
      resource: "tests",
      verbs: ["get", "create"],
    },
    {
      apiGroup: "apps",
      namespace: "test",
      resource: "deployments",
      verbs: ["list", "watch"],
    },
  ];
  const wrapper = shallow(
    <PermissionsErrorAlert roles={roles} action="test" namespace="test" rawMessage="" />,
  );
  expect(wrapper.html()).toContain("Ask your administrator for the following RBAC roles:");
  expect(wrapper.find(PermissionsListItem)).toHaveLength(2);
});

it("renders a link to access control documentation", () => {
  const roles: IRBACRole[] = [];
  const wrapper = shallow(
    <PermissionsErrorAlert roles={roles} action="test" namespace="test" rawMessage="" />,
  );
  expect(wrapper.html()).toMatch(
    /See the documentation for more info on.*access control in Kubeapps./,
  );
  expect(wrapper.html()).toContain(
    '<a href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md" target="_blank" rel="noopener noreferrer">',
  );
  expect(wrapper.html()).not.toContain(shallow(genericMessage).html());
});

it("renders the raw message if there are no roles defined", () => {
  const roles: IRBACRole[] = [];
  const wrapper = shallow(
    <PermissionsErrorAlert roles={roles} action="test" namespace="test" rawMessage="boom!" />,
  );
  expect(wrapper.html()).toMatch(/boom!/);
  expect(wrapper.html()).not.toContain(shallow(genericMessage).html());
});
