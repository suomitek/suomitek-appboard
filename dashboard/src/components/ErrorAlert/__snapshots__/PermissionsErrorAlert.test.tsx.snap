// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an error message for the action 1`] = `
<UnexpectedErrorPage
  icon={[Function]}
  showGenericMessage={false}
  title={
    <span>
      You don't have sufficient permissions to 
      unit-test
       
      <React.Fragment>
        in 
        <span>
          the 
          <code>
            test
          </code>
           namespace
        </span>
      </React.Fragment>
    </span>
  }
>
  <div>
    <p>
      The following error was returned:
    </p>
    <div
      className="error__content"
    >
      <section
        className="Terminal terminal__error elevation-1 type-color-white error__text"
      />
    </div>
    <p>
      See the documentation for more info on
       
      <a
        href="https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md"
        rel="noopener noreferrer"
        target="_blank"
      >
        access control in Kubeapps
      </a>
      .
    </p>
  </div>
</UnexpectedErrorPage>
`;
